src/Storages/MergeTree/CommonANNIndexes.h

```diff
@@ -0,0 +1,224 @@
+    {
+        Int64 length;
+    };
+
+    using ANNExprOpt = std::optional<ANNExpression>;
+    using LimitExprOpt = std::optional<LimitExpression>;
+
+    struct RPNElement
+    {
+        enum Function
+        {
+            // DistanceFunctions
+            FUNCTION_DISTANCE,
+
+            //tuple(0.1, ..., 0.1)
+            FUNCTION_TUPLE,
+
+            // Operators <, >, <=, >=
+            FUNCTION_COMPARISON,
+
+            // Numeric float value
+            FUNCTION_FLOAT_LITERAL,
+
+            // Numeric int value
+            FUNCTION_INT_LITERAL,
+
+            // Column identifier
+            FUNCTION_IDENTIFIER,
+
+            // Unknown, can be any value
+            FUNCTION_UNKNOWN,
+
+            // (0.1, ...., 0.1) vector without word 'tuple'
+            FUNCTION_LITERAL_TUPLE,
+        };
+
+        explicit RPNElement(Function function_ = FUNCTION_UNKNOWN)
+        : function(function_), func_name("Unknown"), float_literal(std::nullopt), identifier(std::nullopt) {}
+
+        Function function;
+        String func_name;
+
+        std::optional<float> float_literal;
+        std::optional<String> identifier;
+        std::optional<int64_t> int_literal{std::nullopt};
+        std::optional<Tuple> tuple_literal{std::nullopt};
+
+        UInt32 dim{0};
+    };
+
+    using RPN = std::vector<RPNElement>;
+
+    void buildRPN(const SelectQueryInfo & query);
+
+    // Util functions for the traversal of AST, parses AST and builds rpn
+    void traverseAST(const ASTPtr & node, RPN & rpn);
+    // Return true if we can identify our node type
+    bool traverseAtomAST(const ASTPtr & node, RPNElement & out);
+    // Checks if the AST stores ConstType expression
+    bool tryCastToConstType(const ASTPtr & node, RPNElement & out);
+    // Traverses the AST of ORDERBY section
+    void traverseOrderByAST(const ASTPtr & node, RPN & rpn);
+
+    // Checks that at least one rpn is matching for index
+    // New RPNs for other query types can be added here
+    bool matchAllRPNS();
+
+    // Returns true and stores ANNExpr if the query has valid WHERE section
+    static bool matchRPNWhere(RPN & rpn, ANNExpression & expr);
+
+    // Returns true and stores ANNExpr if the query has valid ORDERBY section
+    static bool matchRPNOrderBy(RPN & rpn, ANNExpression & expr);
+
+    // Returns true and stores Length if we have valid LIMIT clause in query
+    static bool matchRPNLimit(RPN & rpn, LimitExpression & expr);
+
+    /* Matches dist function, target vector, column name */
+    static bool matchMainParts(RPN::iterator & iter, RPN::iterator & end, ANNExpression & expr, bool & identifier_found);
+
+    // Util methods
+    static void panicIfWrongBuiltRPN [[noreturn]] ();
+    static String getIdentifierOrPanic(RPN::iterator& iter);
+    // Gets float or int from AST node
+    static float getFloatOrIntLiteralOrPanic(RPN::iterator& iter);
+
+
+    // RPN-s for different sections of the query
+    RPN rpn_prewhere_clause;
```

Comment:
As far as I see, the rpn_* members are only created and used temporarily during initialization of a Condition object. If so, we should not store them as members.